Mathlib/Tactic/Says.lean

```diff
@@ -0,0 +1,88 @@
+/-
+Copyright (c) 2023 Kim Liesinger. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kim Liesinger
+-/
+import Std.Data.String.Basic
+import Std.Tactic.GuardMsgs
+
+/-!
+# The `says` tactic combinator.
+
+If you write `X says`, where `X` is a tactic that produces a "Try this: Y" message,
+then you will get a message "Try this: X says Y".
+Once you've clicked to replace `X says` with `X says Y`,
+afterwards `X says Y` will only run `Y`.
+
+The typical usage case is:
+```
+simp? [X] says simp only [X, Y, Z]
+```
+-/
+
+open Lean Elab Tactic
+open Std.Tactic.TryThis
+
+namespace Mathlib.Tactic.Says
+
+register_option says.verify : Bool :=
+  { defValue := false
+    group := "says"
+    descr := "For every appearance of the `X says Y` combinator," ++
+      " re-verify that running `X` produces `Try this: Y`." }
+
+/-- Run `evalTactic`, capturing any new messages.-/
+def evalTacticCapturingMessages (tac : TSyntax `tactic) : TacticM MessageLog := do
+  let initMsgs ← modifyGetThe Core.State fun st => (st.messages, { st with messages := {} })
+  evalTactic tac
+  let msgs := (← getThe Core.State).messages
+  modifyThe Core.State fun st => { st with messages := initMsgs }
+  return msgs
```

Comment:
This function currently drops all mesasges if `evalTactic` throws an error. I didn't test it, but it seems to me that what we want is that in the failure case, we restore all the messages, add the new messages, and propagate the error. I didn't test it, but something like this:
```suggestion
  let mut msgs ← modifyGetThe Core.State fun st => (st.messages, { st with messages := {} })
  try
    evalTactic tac
    return (← getThe Core.State).messages
  catch e =>
    msgs := msgs ++ (← getThe Core.State).messages
    throw e
  finally
    modifyThe Core.State fun st => { st with messages := msgs }
```

Reply:
Implemented! Also did some further refactoring, adding `evalTacticCapturingInfo` and `evalTacticCapturingTryThis`.